javascript/computer_science/hash_map_data_structure.md

```diff
@@ -0,0 +1,204 @@
+
+1. Pass "Fred" into the hash function to get the hash code which is `508`
+1. Find the bucket at index `508`
+1. Store the key value pair in that bucket.
+
+This is an oversimplified explanation; we'll discuss more internal mechanics later in the lesson.
+
+To get a value using a key, we put each entry inside a bucket as a `Node` item, which hold both the key and the value. To retrieve the value we hash a key, we find the bucket number, if the bucket is not empty then we go to that bucket we compare if the Node's key is the same key that is already in the bucket, if it is then we retrieve the Node's value otherwise we return null.
+
+Maybe you are wondering, why are we comparing the keys if we already found the index of that bucket? That is because remember, hash code is just the location, different keys might generate the same hash code. We need to make sure the key is the same by comparing both keys that are inside the bucket.
+
+This is it, making this will result in a hash table with `has`, `set` and `get`.
+
+What if we found the hash code, but also the key value is the same as what we already have in the bucket. We check if it's the same item by comparing the keys, then we overwrite the value with our new value. This is how we can only have unique values inside a `Set`, `Set` is similar to a hash map but the key difference (pun intended) is that a `Set` will have Nodes with only keys and no values.
+
+#### Insertion order is not maintained
+
+A hash map does not guarantee insertion order when you iterate over it. Hash codes translation to indexes does not go in linear fashion from first to last index, and rather more chaotic of where the index going to be no matter the insert order. That means if you are to retrieve the array of keys and values to iterate over them, then they will not be in order of when you inserted them.
+
+Some libraries implement hash tables with insertion order in mind such as JavaScript's own `Map`. For this project however we will be implementing an unordered hash table.
+Example: if we insert the values `Mao`, `Zach`, `Xari` in this order, we may get back `["Zach", "Mao", "Xari"]` when we call an iterator.
+
+If iterating over the hash map frequently is your goal, then this data structure is not the right choice for the job, a simple array would be better.
+
+### Collisions
+
+We have another problem that we need to deal with: collisions. A collision occurs when two different keys generate the exact same hash code. Because they have the same hash code, they will land in the same bucket.
+
+Let's take an example: hashing the name `"Sara"` and the name `"raSa"` will generate the same hash code. That is because the letters in both names are the same, just arranged differently.
+We can rework our `stringToNumber` function so that it can give us unique hash codes which depends on where the letter appear in the name using an algorithm.
```

Comment:
```suggestion
We can rework our `stringToNumber` function so that it can give us unique hash codes that depend on where the letters appear in the name using an algorithm.
```